[CR] github: don't leave behind source file when moving

felliott opened the pull request:
When moving files from a non-root folder in github, the source file was  being left behind.  This was because the tree entry for the source directory was left in the modified master tree, where it still contained references to the old blob/tree.  The solution was to remove all tree entries from the master tree and to allow github to infer the new tree structure from the path names of the remaining blobs.

I also included a fix for folder move/copy actions returning the wrong status code, since the changes are very closely related.

Fixes: [#OSF-4792], [#OSF-4870]

felliott commented:
I don't have proper unit tests for this, b/c it's a lot to mock up, but I do have a Paw session I can run for you to verify.  I am also planning on adding a Runscope regression test for this.

chrisseto commented on waterbutler/providers/github/provider.py:
```diff
@@ -567,34 +567,48 @@ def _update_ref(self, sha, ref='master'):
 
-        if target is None or (src_path.is_dir and target['type'] != 'tree'):
+        if len(blobs) == 0:
             raise exceptions.NotFoundError(str(src_path))
```
This would mean it is impossible to have an empty folder in git, correct?
Which is true, just wanted to make sure everyone is on the same page.

felliott replied:
Yep.

chrisseto commented on waterbutler/providers/github/provider.py:
```diff
@@ -567,34 +567,48 @@ def _update_ref(self, sha, ref='master'):
 
+        # github infers tree contents from blob paths
+        # see: http://www.levibotelho.com/development/commit-a-file-with-the-github-api/
+        tree['tree'][:] = [item for item in tree['tree'] if item['type'] != 'tree']
```
What is the purpose of the `[:]` here? I've never seen that syntax before

felliott replied:
I.... don't know.  It might of been a copy-and-paste artifact that accidentally worked?  I'll delete it.

chrisseto commented on waterbutler/providers/github/provider.py:
```diff
@@ -628,7 +642,7 @@ def _do_intra_move_or_copy(self, src_path, dest_path, is_copy):
 
         if dest_path.is_file:
-            return GitHubFileTreeMetadata(target, commit=commit), not exists
+            return GitHubFileTreeMetadata(blobs[0], commit=commit), not exists
```
`blobs` should always have a length of 1 here?
If so could you add an assertion?

felliott replied:
Will do.

chrisseto commented on waterbutler/providers/github/provider.py:
```diff
@@ -567,34 +567,48 @@ def _update_ref(self, sha, ref='master'):
 
-        target['path'] = target['path'].replace(src_path.path.strip('/'), dest_path.path.strip('/'), 1)
+        for blob in blobs:
+            blob['path'] = blob['path'].replace(src_path.path, dest_path.path, 1)
```
What's the purpose of doing this? it looks like `blobs` is only referenced once after this?

felliott replied:
`blobs` are references to the blob entries from the original tree.  By modifying them we modify the tree structure, yes?

chrisseto commented on waterbutler/providers/github/provider.py:
```diff
@@ -646,11 +661,13 @@ def _do_intra_move_or_copy(self, src_path, dest_path, is_copy):
 
-        for item in keep:
-            item['path'] = item['path'].replace(src_path.path, dest_path.path, 1)
+        tree = yield from self._fetch_tree(orig_tree['sha'])
```
Any chance we can avoid making this request?

felliott replied:
Done, with the rest of the code simplified in the process.

chrisseto commented on waterbutler/providers/github/provider.py:
```diff
@@ -577,35 +577,43 @@ def _update_ref(self, sha, ref='master'):
+        exists = any(x['path'] == dest_path.path.rstrip('/') for x in tree['tree'])
 
-        target, keep = None, []
+        blobs = [item for item in tree['tree'] if item['path'].startswith(src_path.path)]
```
Should this be `rstrip`ed?

felliott replied:
No.  We want to identify all children of the path.  If we rstrip, we could end up pulling out paths with similar names..      mv foo/   could end up moving foobar.txt.

...dang, but now I think there's a bug when handling files.  I might need to add another check.

chrisseto replied:
Something like this, maybe?
```python
[
    item
    for item in tree['tree'] 
    if item['path'].startswith(src_path.path) 
    or item[path'] == src_path.path.rstrip('/')
]
```

chrisseto commented on waterbutler/providers/github/provider.py:
```diff
@@ -646,11 +655,13 @@ def _do_intra_move_or_copy(self, src_path, dest_path, is_copy):
         folder.children = []
 
-        for item in keep:
+        for item in blob_copies:
```
Couldn't you just iterate of the `blobs` list as that one already has the paths adjusted?
I remember commenting about this previously and I'm very sorry if the two comments are conflicting.

felliott replied:
I remember I had a specific reason for doing this, but I'm blanking on what it was.  I think I was worried about having the trees removed, but know that I look at it, that shouldn't happen.  I'll reinvestigate (and add a comment).

chrisseto commented:
Looking good, 2 final comments

felliott commented:
Updated.  I couldn't remember my rationale for the second comment about `blob_copies`, so I've done away with that.  For the blob extraction, I added an extra conditional to make sure that `mv /foo /A/` doesn't try to move `/foobar` as well.

chrisseto commented on waterbutler/providers/github/provider.py:
```diff
@@ -577,35 +577,46 @@ def _update_ref(self, sha, ref='master'):
 
+        # github infers tree contents from blob paths
+        # see: http://www.levibotelho.com/development/commit-a-file-with-the-github-api/
+        tree['tree'] = [item for item in tree['tree'] if item['type'] != 'tree']
```
Have you tried without this line?
Just wondering if github had fixed this issue.

felliott replied:
> In waterbutler/providers/github/provider.py <https://github.com/CenterForOpenScience/waterbutler/pull/83#discussion_r43536590>:
> >  
> > +        # github infers tree contents from blob paths
> > +        # see: http://www.levibotelho.com/development/commit-a-file-with-the-github-api/
> > +        tree['tree'] = [item for item in tree['tree'] if item['type'] != 'tree']
> Have you tried without this line?
> Just wondering if github had fixed this issue.
> 

Just did, and it still fails.  I don’t think this is a bug on their end, just a side effect of the git model

chrisseto commented:
LGTM
@felliott, are you ready for this to be merged?

cwisecarver commented:
I just went through the code as a CR2. Nothing jumped out at me but I also couldn't add github to my local osf. I was having some sort of problem around the oauth exchange. I'm happy to take another look if a helpful @felliott or @chrisseto gives me some protips.

felliott commented:
> On Oct 30, 2015, at 2:32 PM, Chris Seto <notifications@github.com> wrote:
> 
> LGTM
> @felliott <https://github.com/felliott>, are you ready for this to be merged?
> 
> 

Sure, but let me see if we can get @cwisecarver’s oauth working so he can put it through its paces.  I’ll send him a message on Flowdock.

cwisecarver commented:
@felliott @chrisseto Everything checks out in practice. Correct status codes coming back from move operations and overwrite operations. Things get moved inside the root, outside the root, from the root to outside the root and vice versa. :+1: for CR2.

chrisseto commented:
Thanks for the update @cwisecarver, I'm going to hold off until merging until some of QA's time is freed up
